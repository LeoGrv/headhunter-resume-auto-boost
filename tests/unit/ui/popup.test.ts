import { jest } from '@jest/globals';
import fs from 'fs';
import path from 'path';

// Mock Chrome APIs for popup testing
const mockChrome: any = {
  runtime: {
    sendMessage: jest.fn(),
    getManifest: jest.fn(() => ({
      version: '1.0.0',
      name: 'HeadHunter Resume Auto-Boost',
    })),
    id: 'test-extension-id',
  },
  storage: {
    local: {
      get: jest.fn(),
      set: jest.fn(),
      remove: jest.fn(),
      clear: jest.fn(),
    },
  },
  tabs: {
    query: jest.fn(),
    get: jest.fn(),
  },
};

// Set up global chrome object
(global as any).chrome = mockChrome;

// Mock DOM environment
const { JSDOM } = require('jsdom');

describe('Popup UI Snapshot Tests', () => {
  let dom: any;
  let document: Document;
  let window: Window;

  beforeEach(() => {
    // Reset all mocks
    jest.clearAllMocks();

    // Mock successful storage operations
    mockChrome.storage.local.get.mockImplementation(
      (keys: any, callback: any) => {
        const result = Array.isArray(keys)
          ? keys.reduce((acc: any, key: any) => ({ ...acc, [key]: null }), {})
          : { [keys]: null };
        if (callback) callback(result);
        return Promise.resolve(result);
      }
    );

    mockChrome.storage.local.set.mockImplementation(
      (items: any, callback: any) => {
        if (callback) callback();
        return Promise.resolve();
      }
    );

    // Read the actual popup HTML
    const popupHtmlPath = path.join(__dirname, '../../../src/popup/popup.html');
    const popupHtml = fs.readFileSync(popupHtmlPath, 'utf8');

    // Create JSDOM instance
    dom = new JSDOM(popupHtml, {
      url: 'chrome-extension://test/popup.html',
      pretendToBeVisual: true,
      resources: 'usable',
    });

    document = dom.window.document;
    window = dom.window;

    // Set up global DOM objects
    (global as any).document = document;
    (global as any).window = window;
    (global as any).HTMLElement = (window as any).HTMLElement;
    (global as any).Element = (window as any).Element;
  });

  afterEach(() => {
    if (dom) {
      dom.window.close();
    }
  });

  describe('Initial HTML Structure', () => {
    it('should match popup HTML snapshot', () => {
      expect(document.documentElement.outerHTML).toMatchSnapshot();
    });

    it('should have correct header structure', () => {
      const header = document.querySelector('header');
      expect(header?.outerHTML).toMatchSnapshot();
    });

    it('should have correct main content structure', () => {
      const main = document.querySelector('main');
      expect(main?.outerHTML).toMatchSnapshot();
    });

    it('should have tabs section', () => {
      const tabsSection = document.querySelector('.tabs-section');
      expect(tabsSection?.outerHTML).toMatchSnapshot();
    });

    it('should have controls section', () => {
      const controlsSection = document.querySelector('.controls-section');
      expect(controlsSection?.outerHTML).toMatchSnapshot();
    });

    it('should have logs section', () => {
      const logsSection = document.querySelector('.logs-section');
      expect(logsSection?.outerHTML).toMatchSnapshot();
    });
  });

  describe('Dynamic Content States', () => {
    it('should match snapshot with loading state', () => {
      const statusText = document.querySelector('.status-text');
      if (statusText) {
        statusText.textContent = 'Loading...';
      }

      const statusDot = document.querySelector('.status-dot');
      if (statusDot) {
        statusDot.className = 'status-dot loading';
      }

      expect(document.querySelector('header')?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with active state', () => {
      const statusText = document.querySelector('.status-text');
      if (statusText) {
        statusText.textContent = 'Active (3 tabs)';
      }

      const statusDot = document.querySelector('.status-dot');
      if (statusDot) {
        statusDot.className = 'status-dot active';
      }

      expect(document.querySelector('header')?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with paused state', () => {
      const statusText = document.querySelector('.status-text');
      if (statusText) {
        statusText.textContent = 'Paused';
      }

      const statusDot = document.querySelector('.status-dot');
      if (statusDot) {
        statusDot.className = 'status-dot paused';
      }

      expect(document.querySelector('header')?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with sample tabs', () => {
      const tabsList = document.querySelector('.tabs-list');
      if (tabsList) {
        tabsList.innerHTML = `
          <div class="tab-item active">
            <div class="tab-info">
              <span class="tab-title">Senior Frontend Developer</span>
              <span class="tab-url">hh.ru/resume/123456</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status">Next boost: 15:30</span>
              <button class="btn-small pause">‚è∏</button>
            </div>
          </div>
          <div class="tab-item paused">
            <div class="tab-info">
              <span class="tab-title">Full Stack Developer</span>
              <span class="tab-url">hh.ru/resume/789012</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status">Paused</span>
              <button class="btn-small resume">‚ñ∂</button>
            </div>
          </div>
        `;
      }

      expect(
        document.querySelector('.tabs-section')?.outerHTML
      ).toMatchSnapshot();
    });

    it('should match snapshot with sample logs', () => {
      const logsList = document.querySelector('.logs-list');
      if (logsList) {
        logsList.innerHTML = `
          <div class="log-entry success">
            <span class="log-time">14:25:30</span>
            <span class="log-message">Resume boosted successfully</span>
            <span class="log-tab">Tab 1</span>
          </div>
          <div class="log-entry info">
            <span class="log-time">14:20:15</span>
            <span class="log-message">Timer started for new tab</span>
            <span class="log-tab">Tab 2</span>
          </div>
          <div class="log-entry warning">
            <span class="log-time">14:15:45</span>
            <span class="log-message">Boost limit reached for today</span>
            <span class="log-tab">Tab 1</span>
          </div>
        `;
      }

      expect(
        document.querySelector('.logs-section')?.outerHTML
      ).toMatchSnapshot();
    });
  });

  describe('Button States', () => {
    it('should match snapshot with pause button active', () => {
      const pauseButton = document.querySelector('#global-pause');
      if (pauseButton) {
        pauseButton.textContent = '‚è∏ Pause All';
        pauseButton.className = 'btn btn-secondary';
      }

      expect(pauseButton?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with resume button active', () => {
      const pauseButton = document.querySelector('#global-pause');
      if (pauseButton) {
        pauseButton.textContent = '‚ñ∂ Resume All';
        pauseButton.className = 'btn btn-primary';
      }

      expect(pauseButton?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with settings button', () => {
      const settingsButton = document.querySelector('#settings');
      expect(settingsButton?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with clear logs button', () => {
      const clearLogsButton = document.querySelector('#clear-logs');
      expect(clearLogsButton?.outerHTML).toMatchSnapshot();
    });
  });

  describe('Empty States', () => {
    it('should match snapshot with no tabs', () => {
      const tabsList = document.querySelector('.tabs-list');
      if (tabsList) {
        tabsList.innerHTML =
          '<div class="empty-state">No resume tabs detected</div>';
      }

      expect(
        document.querySelector('.tabs-section')?.outerHTML
      ).toMatchSnapshot();
    });

    it('should match snapshot with no logs', () => {
      const logsList = document.querySelector('.logs-list');
      if (logsList) {
        logsList.innerHTML =
          '<div class="empty-state">No recent activity</div>';
      }

      expect(
        document.querySelector('.logs-section')?.outerHTML
      ).toMatchSnapshot();
    });
  });

  describe('Error States', () => {
    it('should match snapshot with error state', () => {
      const statusText = document.querySelector('.status-text');
      if (statusText) {
        statusText.textContent = 'Error: Connection failed';
      }

      const statusDot = document.querySelector('.status-dot');
      if (statusDot) {
        statusDot.className = 'status-dot error';
      }

      expect(document.querySelector('header')?.outerHTML).toMatchSnapshot();
    });

    it('should match snapshot with tab error', () => {
      const tabsList = document.querySelector('.tabs-list');
      if (tabsList) {
        tabsList.innerHTML = `
          <div class="tab-item error">
            <div class="tab-info">
              <span class="tab-title">Resume Tab (Error)</span>
              <span class="tab-url">hh.ru/resume/error</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status error">Boost failed</span>
              <button class="btn-small retry">üîÑ</button>
            </div>
          </div>
        `;
      }

      expect(
        document.querySelector('.tabs-section')?.outerHTML
      ).toMatchSnapshot();
    });
  });
});
