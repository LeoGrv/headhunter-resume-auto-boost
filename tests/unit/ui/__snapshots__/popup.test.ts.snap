// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup UI Snapshot Tests Button States should match snapshot with clear logs button 1`] = `"<button id="clear-logs" class="btn btn-small">Clear Logs</button>"`;

exports[`Popup UI Snapshot Tests Button States should match snapshot with pause button active 1`] = `"<button id="global-pause" class="btn btn-secondary">‚è∏ Pause All</button>"`;

exports[`Popup UI Snapshot Tests Button States should match snapshot with resume button active 1`] = `"<button id="global-pause" class="btn btn-primary">‚ñ∂ Resume All</button>"`;

exports[`Popup UI Snapshot Tests Button States should match snapshot with settings button 1`] = `"<button id="settings" class="btn btn-primary">‚öôÔ∏è Settings</button>"`;

exports[`Popup UI Snapshot Tests Dynamic Content States should match snapshot with active state 1`] = `
"<header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot active"></span>
                <span class="status-text">Active (3 tabs)</span>
            </div>
        </header>"
`;

exports[`Popup UI Snapshot Tests Dynamic Content States should match snapshot with loading state 1`] = `
"<header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot loading"></span>
                <span class="status-text">Loading...</span>
            </div>
        </header>"
`;

exports[`Popup UI Snapshot Tests Dynamic Content States should match snapshot with paused state 1`] = `
"<header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot paused"></span>
                <span class="status-text">Paused</span>
            </div>
        </header>"
`;

exports[`Popup UI Snapshot Tests Dynamic Content States should match snapshot with sample logs 1`] = `
"<div class="logs-section">
                <h2>Recent Activity</h2>
                <div class="logs-list">
          <div class="log-entry success">
            <span class="log-time">14:25:30</span>
            <span class="log-message">Resume boosted successfully</span>
            <span class="log-tab">Tab 1</span>
          </div>
          <div class="log-entry info">
            <span class="log-time">14:20:15</span>
            <span class="log-message">Timer started for new tab</span>
            <span class="log-tab">Tab 2</span>
          </div>
          <div class="log-entry warning">
            <span class="log-time">14:15:45</span>
            <span class="log-message">Boost limit reached for today</span>
            <span class="log-tab">Tab 1</span>
          </div>
        </div>
                <button id="clear-logs" class="btn btn-small">Clear Logs</button>
            </div>"
`;

exports[`Popup UI Snapshot Tests Dynamic Content States should match snapshot with sample tabs 1`] = `
"<div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list">
          <div class="tab-item active">
            <div class="tab-info">
              <span class="tab-title">Senior Frontend Developer</span>
              <span class="tab-url">hh.ru/resume/123456</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status">Next boost: 15:30</span>
              <button class="btn-small pause">‚è∏</button>
            </div>
          </div>
          <div class="tab-item paused">
            <div class="tab-info">
              <span class="tab-title">Full Stack Developer</span>
              <span class="tab-url">hh.ru/resume/789012</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status">Paused</span>
              <button class="btn-small resume">‚ñ∂</button>
            </div>
          </div>
        </div>
            </div>"
`;

exports[`Popup UI Snapshot Tests Empty States should match snapshot with no logs 1`] = `
"<div class="logs-section">
                <h2>Recent Activity</h2>
                <div class="logs-list"><div class="empty-state">No recent activity</div></div>
                <button id="clear-logs" class="btn btn-small">Clear Logs</button>
            </div>"
`;

exports[`Popup UI Snapshot Tests Empty States should match snapshot with no tabs 1`] = `
"<div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list"><div class="empty-state">No resume tabs detected</div></div>
            </div>"
`;

exports[`Popup UI Snapshot Tests Error States should match snapshot with error state 1`] = `
"<header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot error"></span>
                <span class="status-text">Error: Connection failed</span>
            </div>
        </header>"
`;

exports[`Popup UI Snapshot Tests Error States should match snapshot with tab error 1`] = `
"<div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list">
          <div class="tab-item error">
            <div class="tab-info">
              <span class="tab-title">Resume Tab (Error)</span>
              <span class="tab-url">hh.ru/resume/error</span>
            </div>
            <div class="tab-controls">
              <span class="timer-status error">Boost failed</span>
              <button class="btn-small retry">üîÑ</button>
            </div>
          </div>
        </div>
            </div>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should have controls section 1`] = `
"<div class="controls-section">
                <h2>Controls</h2>
                <div class="global-controls">
                    <button id="global-pause" class="btn btn-secondary">‚è∏ Pause All</button>
                    <button id="settings" class="btn btn-primary">‚öôÔ∏è Settings</button>
                </div>
            </div>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should have correct header structure 1`] = `
"<header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot"></span>
                <span class="status-text">Loading...</span>
            </div>
        </header>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should have correct main content structure 1`] = `
"<main>
            <div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
            </div>
            
            <div class="controls-section">
                <h2>Controls</h2>
                <div class="global-controls">
                    <button id="global-pause" class="btn btn-secondary">‚è∏ Pause All</button>
                    <button id="settings" class="btn btn-primary">‚öôÔ∏è Settings</button>
                </div>
            </div>
            
            <div class="logs-section">
                <h2>Recent Activity</h2>
                <div class="logs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
                <button id="clear-logs" class="btn btn-small">Clear Logs</button>
            </div>
        </main>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should have logs section 1`] = `
"<div class="logs-section">
                <h2>Recent Activity</h2>
                <div class="logs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
                <button id="clear-logs" class="btn btn-small">Clear Logs</button>
            </div>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should have tabs section 1`] = `
"<div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
            </div>"
`;

exports[`Popup UI Snapshot Tests Initial HTML Structure should match popup HTML snapshot 1`] = `
"<html lang="en"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>HeadHunter Resume Auto-Boost</title>
    <link rel="stylesheet" href="popup.css">
</head>
<body>
    <div class="container">
        <header>
            <h1>HH Resume Boost</h1>
            <div class="status-indicator">
                <span class="status-dot"></span>
                <span class="status-text">Loading...</span>
            </div>
        </header>
        
        <main>
            <div class="tabs-section">
                <h2>Resume Tabs</h2>
                <div class="tabs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
            </div>
            
            <div class="controls-section">
                <h2>Controls</h2>
                <div class="global-controls">
                    <button id="global-pause" class="btn btn-secondary">‚è∏ Pause All</button>
                    <button id="settings" class="btn btn-primary">‚öôÔ∏è Settings</button>
                </div>
            </div>
            
            <div class="logs-section">
                <h2>Recent Activity</h2>
                <div class="logs-list">
                    <!-- Dynamic content will be inserted here -->
                </div>
                <button id="clear-logs" class="btn btn-small">Clear Logs</button>
            </div>
        </main>
    </div>
    
    <script src="popup.js"></script>

 </body></html>"
`;
